sszgen/hash.go

```diff
@@ -95,78 +95,112 @@ func (v *Value) hashRoots(isList bool, elem Type) string {
+// takes a "name" param so that the variable name can be replaced with a local name
+// ie within a for loop for a list, the we want to refer to "elem" w/o a receiver variable
+// when not specified, name will be set to "::." + v.name. In the final templating pass,
+// the output formatter replaces all instances of "::" with the receiver variable for the container.
+func (v *Value) hashTreeRoot(name string, appendBytes bool) string {
+	if name == "" {
+		name = "::." + v.name
+	}
 	switch v.t {
 	case TypeContainer, TypeReference:
 		return v.hashTreeRootContainer(false)
 
 	case TypeBytes:
-		// There are only fixed []byte
-		name := v.name
 		if v.c {
 			name += "[:]"
 		}
-
-		tmpl := `{{.validate}}hh.PutBytes(::.{{.name}})`
-		return execTmpl(tmpl, map[string]interface{}{
-			"validate": v.validate(),
-			"name":     name,
-			"size":     v.s,
-		})
+		if v.isFixed() {
+			tmpl := `{{.validate}}hh.PutBytes({{.name}})`
+			return execTmpl(tmpl, map[string]interface{}{
+				"validate": v.validate(),
+				"name":     name,
+				"size":     v.s,
+			})
+		} else {
+			// dynamic bytes require special handling, need length mixed in
+			hMethod := "PutBytes"
+			if appendBytes {
+				hMethod = "AppendBytes32"
+			}
+			tmpl := `{
+	elemIndx := hh.Index()
+	byteLen := uint64(len({{.name}}))
+	if byteLen > {{.maxLen}} {
+		err = ssz.ErrIncorrectListSize
+		return
+    }
+	hh.{{.hashMethod}}({{.name}})
+	hh.MerkleizeWithMixin(elemIndx, byteLen, ({{.maxLen}}+31)/32)
+}`
+			return execTmpl(tmpl, map[string]interface{}{
+				"hashMethod": hMethod,
+				"name": name,
+				"maxLen": v.m,
+			})
+		}
 
 	case TypeUint:
-		var name string
 		if v.ref != "" || v.obj != "" {
 			// alias to Uint64
-			name = fmt.Sprintf("uint64(::.%s)", v.name)
-		} else {
-			name = "::." + v.name
+			name = fmt.Sprintf("uint64(%s)", name)
 		}
-		bitLen := v.n * 8
+		bitLen := v.fixedSize() * 8
 		return fmt.Sprintf("hh.PutUint%d(%s)", bitLen, name)
 
 	case TypeBitList:
-		tmpl := `if len(::.{{.name}}) == 0 {
+		tmpl := `if len({{.name}}) == 0 {
 			err = ssz.ErrEmptyBitlist
 			return
 		}
-		hh.PutBitlist(::.{{.name}}, {{.size}})
+		hh.PutBitlist({{.name}}, {{.size}})
 		`
 		return execTmpl(tmpl, map[string]interface{}{
-			"name": v.name,
+			"name": name,
 			"size": v.m,
 		})
 
 	case TypeBool:
-		return fmt.Sprintf("hh.PutBool(::.%s)", v.name)
+		return fmt.Sprintf("hh.PutBool(%s)", name)
 
 	case TypeVector:
 		return v.hashRoots(false, v.e.t)
 
 	case TypeList:
 		if v.e.isFixed() {
 			if v.e.t == TypeUint || v.e.t == TypeBytes {
-				// return hashBasicSlice(v)
 				return v.hashRoots(true, v.e.t)
 			}
 		}
+
 		tmpl := `{
 			subIndx := hh.Index()
-			num := uint64(len(::.{{.name}}))
+			num := uint64(len({{.name}}))
 			if num > {{.num}} {
 				err = ssz.ErrIncorrectListSize
 				return
 			}
-			for i := uint64(0); i < num; i++ {
-				if err = ::.{{.name}}[i].HashTreeRootWith(hh); err != nil {
-					return
-				}
+			for _, elem := range {{.name}} {
```

Comment:
This change results in changes to codegen beyond the scope of ByteLists, but we've reviewed the impact to codegen in prysm and it is benign as far as we're concerned.